sm/tedge_agent/src/operation_logs.rs

```diff
@@ -0,0 +1,199 @@
+const LIST_PREFIX: &'static str = "software-list";
+
+impl OperationLogs {
+    pub fn try_new(log_dir: PathBuf) -> Result<OperationLogs, std::io::Error> {
+        std::fs::create_dir_all(log_dir.clone())?;
+        let operation_logs = OperationLogs {
+            log_dir,
+            max_update_logs: 5,
+            max_list_logs: 1,
+        };
+
+        if let Err(err) = operation_logs.remove_outdated_logs() {
+            log::warn!("Fail to remove the out-dated log files: {}", err);
+        }
+
+        Ok(operation_logs)
+    }
+
+    pub async fn new_log_file(&self, kind: LogKind) -> Result<LogFile, std::io::Error> {
+        if let Err(err) = self.remove_outdated_logs() {
+            log::warn!("Fail to remove the out-dated log files: {}", err);
+        }
+
+        let now = Utc::now();
+        let file_prefix = match kind {
+            LogKind::SoftwareUpdate => UPDATE_PREFIX,
+            LogKind::SoftwareList => LIST_PREFIX,
+        };
+        let file_name = format!(
+            "{}-{}.log",
+            file_prefix,
+            now.to_rfc3339_opts(SecondsFormat::Secs, true)
+        );
+
+        let mut log_file_path = self.log_dir.clone();
+        log_file_path.push(file_name);
+
+        LogFile::try_new(log_file_path).await
+    }
+
+    pub fn remove_outdated_logs(&self) -> Result<(), std::io::Error> {
+        let mut update_logs = BinaryHeap::new();
+        let mut list_logs = BinaryHeap::new();
+        for entry in self.log_dir.read_dir()? {
+            // Ignore entries with errors
+            if let Ok(file) = entry {
+                if let Some(path) = file.path().file_name().and_then(|name| name.to_str()) {
+                    if path.starts_with(UPDATE_PREFIX) {
+                        // The paths are pushed using the reverse alphabetic order
+                        update_logs.push(Reverse(path.to_string()));
+                    } else if path.starts_with(LIST_PREFIX) {
+                        // The paths are pushed using the reverse alphabetic order
+                        list_logs.push(Reverse(path.to_string()));
+                    }
+                }
+            }
+        }
+
+        while update_logs.len() > self.max_update_logs {
+            if let Some(rname) = update_logs.pop() {
+                let name = rname.0;
+                let path = self.log_dir.join(name.clone());
+                if let Err(err) = std::fs::remove_file(&path) {
+                    log::warn!("Fail to remove out-dated log file {} : {}", name, err);
+                }
+            }
+        }
+
+        while list_logs.len() > self.max_list_logs {
+            if let Some(rname) = list_logs.pop() {
+                let name = rname.0;
+                let path = self.log_dir.join(name.clone());
+                if let Err(err) = std::fs::remove_file(&path) {
+                    log::warn!("Fail to remove out-dated log file {} : {}", name, err);
+                }
+            }
+        }
+
```

Comment:
I remember discussing one log per operation, dint get why there are two logs,  update_logs and list_logs here?.

Reply:
Indeed, I decided to also keep a log of the latest software-list operation.